### PR TITLE
Add compile time instruction count metric

### PR DESCRIPTION
**Pull request by laithsakka**
PYTHONPATH=$(pwd) python benchmarks/update_hint_benchmark.py out
as of this diff, compile_time_instruction_count counts the number of instruction from within
convert_frame.compile_inner
```
update_hint_regression,compile_time_instruction_count,10522459165
```
 will add result from CI once populated.                                               
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #133834

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames @rec

### CODE REVIEW

**Comment by laithsakka** [comment]
address lint error

File: benchmarks/dynamo/pr_time_benchmarks/benchmarks/update_hint_benchmark.py [hunk 1]
```diff
@@ -34,12 +34,15 @@ def f(a, b):
...
 
-        f(self.input, self.splits)
+        for i in range(1000):
+            f(self.input, self.splits)
 
 
 def main():
     result_path = sys.argv[1]
-    Benchmark().enable_instruction_count().collect_all().append_results(result_path)
+    Benchmark().enable_instruction_count().enable_compile_time_instruction_count().collect_all().append_results(
```
Comment (ezyang) [inline benchmarks/dynamo/pr_time_benchmarks/benchmarks/update_hint_benchmark.py:43]: Are you really proposing that when we have dozens of benchmark we have to manually update each of them to record new metrics we think we want to record?

File: torch/_dynamo/utils.py [hunk 2]
```diff
@@ -3067,3 +3068,42 @@ def get_user_object_from_id(obj_id):
...
+    @classmethod
+    def value(cls):
+        return cls._counter
+
+
+def maybe_count_instructions(func):
+    if config.record_compile_time_instruction_count:
+
+        @functools.wraps(func)
+        def wrapper_function(*args, **kwargs):
```
Comment (ezyang) [inline torch/_dynamo/utils.py:3101]: No decorator please, it makes the cProfiles worse. Do this as a context manager.

**Review by ezyang — CHANGES_REQUESTED** [review]
see comments

File: benchmarks/dynamo/pr_time_benchmarks/benchmark_base.py [hunk 3]
```diff
@@ -2,47 +2,69 @@
...
     _instruction_count = False
+    _compile_time_instruction_count = False
 
     def enable_instruction_count(self):
         self._instruction_count = True
         return self
 
+    def enable_compile_time_instruction_count(self):
+        self._compile_time_instruction_count = True
+        return self
```
Comment (ezyang) [inline benchmarks/dynamo/pr_time_benchmarks/benchmark_base.py:19]: At this point I think it's good to actually add docs to all of these. In particular, it seems good to carefully define what compile time instruction count means, because, e.g., if I do a forward-backward compile, I'm pretty sure your current formulation doesn't capture backwards compilation.

File: benchmarks/dynamo/pr_time_benchmarks/benchmarks/update_hint_benchmark.py [hunk 4]
```diff
@@ -34,12 +34,15 @@ def f(a, b):
                 torch._check(x <= self.N)
             return a.split(xs)
 
-        f(self.input, self.splits)
+        for i in range(1000):
+            f(self.input, self.splits)
```
Comment (ezyang) [inline benchmarks/dynamo/pr_time_benchmarks/benchmarks/update_hint_benchmark.py:38]: If you want to be in the business of measuring the instruction count of compiled codes, it seems like it would be better to do a measurement here that *excludes* compilation time from the perf measurement. The most typical way to do this would be to do the warmup (forcing compilation) and then do the perf measurement in steady-state. Otherwise, your overall instruction count number is going to bounce up and down, mostly due to changes in compile time, not so useful for finding runtime regressions unless you take care to subtract out the compile time.

File: torch/_dynamo/utils.py [hunk 5]
```diff
@@ -3067,3 +3068,42 @@ def get_user_object_from_id(obj_id):
...
     obj_id = id(obj)
     user_obj_id_to_weakref[obj_id] = weakref.ref(obj)
+
+
+class CompileTimeInstructionCounter:
+    _counter: int = 0
+    _id: Optional[int] = None
+
+    @classmethod
+    def start(cls):
```
Comment (ezyang) [inline torch/_dynamo/utils.py:3078]: nit: -> None on this and all the other class methods

File: benchmarks/dynamo/pr_time_benchmarks/benchmarks/update_hint_benchmark.py (hunk 1 above)
Comment (laithsakka) [reply benchmarks/dynamo/pr_time_benchmarks/benchmarks/update_hint_benchmark.py:43]: i assume those benchmarks are going to be added one by one? and not all metrics need to be enabled on all of them. 
if someone is adding  a dozen of benchmarks and want to enable all metrics on all of them then he can easily create a function (enableAll()).

File: benchmarks/dynamo/pr_time_benchmarks/benchmark_base.py (hunk 3 above)
Comment (laithsakka) [reply benchmarks/dynamo/pr_time_benchmarks/benchmark_base.py:19]: i will be very interesting if it does not because that means we do not capture that in our c_profiles nor out strobelight profiler, because   @maybe_cprofile is only wrapping _compile_inner

File: benchmarks/dynamo/pr_time_benchmarks/benchmarks/update_hint_benchmark.py (hunk 4 above)
Comment (laithsakka) [reply benchmarks/dynamo/pr_time_benchmarks/benchmarks/update_hint_benchmark.py:38]: the instruction count metric, measure the total instruction count for the work function, this could be something that does not even do compilation . (people can add micro benchmarks). 

I see your point about adding another metric which is " instruction count of compiled codes" but that would be a different diff. 

for the time being i can exclude the enable_instruction_count() for the update hint benchmark since what we care about for that one is compile time instruction count. 

I was using that as basic benchmark example, maybe i will add a dummy benchmark were i enable everything on it as example.

File: benchmarks/dynamo/pr_time_benchmarks/benchmark_base.py (hunk 3 above)
Comment (laithsakka) [reply benchmarks/dynamo/pr_time_benchmarks/benchmark_base.py:19]: I will add a documentation though

File: torch/_dynamo/utils.py (hunk 2 above)
Comment (laithsakka) [reply torch/_dynamo/utils.py:3101]: I can do that.

File: benchmarks/dynamo/pr_time_benchmarks/benchmark_base.py (hunk 3 above)
Comment (laithsakka) [reply benchmarks/dynamo/pr_time_benchmarks/benchmark_base.py:19]: is there a specific entry point for the missing compilation part that i can wrap?

**Comment by laithsakka** [comment]
- add type annotation for class methods return 
- document each metric in benchmark base. 
- only collect compile time metric for update_hint benchmark. 
- avoid using decorator.

**Comment by laithsakka** [comment]
- after rebase i noticed the logging to scribe happening from benchmark base, I added a todo, will follow up on this once we update the logger.

**Comment by ezyang** [comment]
I'm on vacation next week, feel free to get someone else to review/approve/land, I don't have any blockers, just haven't done a rereview in time.

File: benchmarks/dynamo/pr_time_benchmarks/benchmark_base.py [hunk 6]
```diff
@@ -51,44 +53,68 @@
...
         pass
 
-    def prepare_once(self):  # noqa: B027
+    def _prepare_once(self):  # noqa: B027
         pass
 
-    def count_instructions(self):
+    def _count_instructions(self):
         print(f"collecting instruction count for {self.name()}")
-        self.prepare_once()
```
Comment (aorenste) [inline benchmarks/dynamo/pr_time_benchmarks/benchmark_base.py:79]: Do we not need this call anymore? Oh - I guess it moved to collect_all()...

File: torch/_dynamo/utils.py [hunk 7]
```diff
@@ -3106,3 +3107,43 @@ def get_user_object_from_id(obj_id):
...
+        return cls._counter
+
+
+class CompileTimeInstructionCollector:
+    def __init__(self):
+        pass
+
+    def __enter__(self):
+        if config.record_compile_time_instruction_count:
+            self.started = True
```
Comment (aorenste) [inline torch/_dynamo/utils.py:3142]: Do we want this to be able to handle recursive calls?
```
with CompileTimeInstructionCollector():
  ...
  with CompileTimeInstructionCollector():
    ...
```
Is so might want a counter instead of a bool...

File: torch/_dynamo/utils.py [hunk 8]
```diff
@@ -56,6 +56,7 @@
 from typing_extensions import Literal, TypeGuard
 
 import torch
+import torch._C._instruction_counter as i_counter
```
Comment (aorenste) [inline torch/_dynamo/utils.py:59]: We only use this twice in this file - I wouldn't bother with such a short name. Maybe just do the import without an alias and call `_instruction_counter.blah` directly?

File: torch/_dynamo/utils.py [hunk 9]
```diff
@@ -3106,3 +3107,43 @@ def get_user_object_from_id(obj_id):
...
+    @classmethod
+    def clear(cls) -> None:
+        cls._counter = 0
+
+    @classmethod
+    def value(cls) -> int:
+        return cls._counter
+
+
+class CompileTimeInstructionCollector:
```
Comment (aorenste) [inline torch/_dynamo/utils.py:3136]: This would probably be easier to read as a contextmanager...

File: torch/_dynamo/utils.py [hunk 10]
```diff
@@ -3106,3 +3107,43 @@ def get_user_object_from_id(obj_id):
...
     obj_id = id(obj)
     user_obj_id_to_weakref[obj_id] = weakref.ref(obj)
+
+
+class CompileTimeInstructionCounter:
+    _counter: int = 0
+    _id: int = -1
+
+    @classmethod
+    def start(cls) -> None:
```
Comment (aorenste) [inline torch/_dynamo/utils.py:3117]: Maybe annotate all these functions?

File: benchmarks/dynamo/pr_time_benchmarks/benchmark_base.py [hunk 11]
```diff
@@ -51,44 +53,68 @@
 
 
 class BenchmarkBase(ABC):
+    # measure total number of instruction spent in _work.
     _instruction_count = False
 
+    # measure total number of instruction spent in convert_frame.compile_inner
+    # TODO is there other parts we need to add ?
+    _compile_time_instruction_count = False
```
Comment (aorenste) [inline benchmarks/dynamo/pr_time_benchmarks/benchmark_base.py:61]: This name is a bit ambiguous (is it a enable? Is it a value?) maybe something like `_enable_compile_time_instruction_count`? (Same issue w/ `_instruction_count` above)

File: benchmarks/dynamo/pr_time_benchmarks/benchmark_base.py [hunk 12]
```diff
@@ -51,44 +53,68 @@
...
+        config.record_compile_time_instruction_count = True
+
+        results = []
+        for i in range(10):
+            self._prepare()
+            CompileTimeInstructionCounter.clear()
+            self._work()
+            count = CompileTimeInstructionCounter.value()
+            print(f"compile time instruction count for iteration {i} is {count}")
+            if i != 0:
```
Comment (aorenste) [inline benchmarks/dynamo/pr_time_benchmarks/benchmark_base.py:112]: Why skip 0? Warmup? Since we're returning `min` warmup shouldn't matter... Please add a comment.

File: benchmarks/dynamo/pr_time_benchmarks/benchmark_base.py (hunk 6 above)
Comment (laithsakka) [reply benchmarks/dynamo/pr_time_benchmarks/benchmark_base.py:79]: yeh once for all metrics not needed

File: benchmarks/dynamo/pr_time_benchmarks/benchmark_base.py (hunk 12 above)
Comment (laithsakka) [reply benchmarks/dynamo/pr_time_benchmarks/benchmark_base.py:112]: yeh it does not really matter we can remove it

File: torch/_dynamo/utils.py (hunk 10 above)
Comment (laithsakka) [reply torch/_dynamo/utils.py:3117]: what do you mean

File: torch/_dynamo/utils.py (hunk 9 above)
Comment (laithsakka) [reply torch/_dynamo/utils.py:3136]: you mean something like
```
from contextlib import contextmanager

@contextmanager
def closing(thing):
    try:
        yield thing
    finally:
        thing.close()
 ```

Comment (aorenste) [reply torch/_dynamo/utils.py:3136]: yes

File: torch/_dynamo/utils.py (hunk 10 above)
Comment (aorenste) [reply torch/_dynamo/utils.py:3117]: Oh - I'm being silly here. @classmethod cls doesn't need annotation - but below the
```
def __init__(self):
```
should be:
```
def __init__(self) -> None:
```
(and enter and exit need type annotations if you don't switch to a contextmanager.

File: torch/_dynamo/utils.py (hunk 7 above)
Comment (laithsakka) [reply torch/_dynamo/utils.py:3142]: right now its not used with recursions but i can make it resilinet to that

File: torch/_dynamo/utils.py [hunk 13]
```diff
@@ -56,6 +56,7 @@
 from typing_extensions import Literal, TypeGuard
 
 import torch
+import torch._C._instruction_counter as _instruction_counter
```
Comment (aorenste) [inline torch/_dynamo/utils.py:59]: nit: this is usually written as:
```
from torch._C import _instruction_counter
```

File: torch/_dynamo/utils.py [hunk 14]
```diff
@@ -3106,3 +3107,39 @@ def get_user_object_from_id(obj_id):
 def store_user_object_weakref(obj):
     obj_id = id(obj)
     user_obj_id_to_weakref[obj_id] = weakref.ref(obj)
+
+
+class CompileTimeInstructionCounter:
```
Comment (aorenste) [inline torch/_dynamo/utils.py:3112]: Why is this "CompileTimeInstructionCounter" and not just "CpuInstructionCounter"? It has no actual relationship to compile time, does it? (Other than we intend to use it for such)

File: benchmarks/dynamo/pr_time_benchmarks/benchmark_base.py [hunk 15]
```diff
@@ -64,29 +75,42 @@ def description(self):
...
-            self.work()
+            self._work()
             count = i_counter.end(id)
             print(f"instruction count for iteration {i} is {count}")
-            if i != 0:
-                results.append(count)
+            results.append(count)
+        return min(results)
+
+    def _count_compile_time_instructions(self):
```
Comment (aorenste) [inline benchmarks/dynamo/pr_time_benchmarks/benchmark_base.py:100]: I'm confused as to the difference between `_count_instructions` and `_count_compile_time_instructions`.  Don't they do the same thing but one uses _instruction_counter directly and the other indirectly?

Comment (laithsakka) [reply benchmarks/dynamo/pr_time_benchmarks/benchmark_base.py:100]: # measure total number of instruction spent in _work.
    _enable_instruction_count = False

    # measure total number of instruction spent in convert_frame.compile_inner
    # TODO is there other parts we need to add ?
    _enable_compile_time_instruction_count = False

Comment (laithsakka) [reply benchmarks/dynamo/pr_time_benchmarks/benchmark_base.py:100]: # measure total number of instruction spent in _work.
 _enable_instruction_count = False 
 
 # measure total number of instruction spent in convert_frame.compile_inner # TODO is there other parts we need to add ? _enable_compile_time_instruction_count = False

File: torch/_dynamo/utils.py (hunk 14 above)
Comment (laithsakka) [reply torch/_dynamo/utils.py:3112]: This should not be used around anything else but counting compile time instructions since its accessed globally 
as CompileTimeInstructionCounter.value()

Comment (laithsakka) [reply torch/_dynamo/utils.py:3112]: if we call it CpuInstructionCounter we can create an a global instance and 
compileTimeInstructionCounter = CpuInstructionCounter()
but i dont see right now a reason for generalizing it

File: benchmarks/dynamo/pr_time_benchmarks/benchmark_base.py (hunk 15 above)
Comment (laithsakka) [reply benchmarks/dynamo/pr_time_benchmarks/benchmark_base.py:100]: if your program  is a micro beanchmark measure some function in isolate. you would use _enable_instruction_count and not _enable_compile_time_instruction_count because
_enable_compile_time_instruction_count would return 0.

File: benchmarks/dynamo/pr_time_benchmarks/benchmark_base.py [hunk 16]
```diff
@@ -64,29 +75,42 @@ def description(self):
...
+        print(f"collecting compile time instruction count for {self.name()}")
+        config.record_compile_time_instruction_count = True
```
Comment (aorenste) [inline benchmarks/dynamo/pr_time_benchmarks/benchmark_base.py:102]: Add a quick comment here noting that because inner_compile (or whatever) calls record() this causes it to actually record the inner_compile ticks and not the whole thing.

**Review by aorenste — APPROVED** [review]
A couple nits but LGTM

File: benchmarks/dynamo/pr_time_benchmarks/benchmark_base.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (laithsakka) [reply benchmarks/dynamo/pr_time_benchmarks/benchmark_base.py:102]: i will add that, but in general if anything throws here the binary will exit anyway.

[the remark this answers is not in the document]
Comment (laithsakka) [reply benchmarks/dynamo/pr_time_benchmarks/benchmark_base.py:102]: actually what are you suggesting what we do in the finally? 
            CompileTimeInstructionCounter.clear() ?
we always clear anyway before work()
